lib/runtime-core/src/instance.rs

```diff
@@ -284,37 +303,45 @@ impl Instance {
     /// # }
     /// ```
     pub fn dyn_func(&self, name: &str) -> ResolveResult<DynFunc> {
-        let export_index =
-            self.module
-                .info
-                .exports
-                .get(name)
-                .ok_or_else(|| ResolveError::ExportNotFound {
-                    name: name.to_string(),
-                })?;
+        let func_index = self.resolve_func(name)?;
 
-        if let ExportIndex::Func(func_index) = export_index {
-            let sig_index = *self
-                .module
-                .info
-                .func_assoc
-                .get(*func_index)
-                .expect("broken invariant, incorrect func index");
-            let signature =
-                SigRegistry.lookup_signature_ref(&self.module.info.signatures[sig_index]);
-
-            Ok(DynFunc {
-                signature,
-                module: &self.module,
-                instance_inner: &self.inner,
-                func_index: *func_index,
-            })
-        } else {
-            Err(ResolveError::ExportWrongType {
-                name: name.to_string(),
-            }
-            .into())
-        }
+        self.dyn_func_with_index(func_index)
+    }
+
+    /// This returns the representation of a function that can be called
+    /// safely.
+    ///
+    /// # Usage:
+    /// ```
+    /// # use wasmer_runtime_core::Instance;
+    /// # use wasmer_runtime_core::error::CallResult;
+    /// # fn call_foo(instance: &mut Instance) -> CallResult<()> {
+    ///
+    /// let index = 0;
+    /// instance
+    ///     .dyn_func_with_index(index)?
+    ///     .call(&[])?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn dyn_func_with_index(&self, func_index: usize) -> ResolveResult<DynFunc> {
+        let func_index = FuncIndex::new(func_index);
+
+        let sig_index = *self
+            .module
+            .info
+            .func_assoc
+            .get(func_index)
+            .expect("broken invariant, incorrect func index");
```

Comment:
> broken invariant, incorrect func index

It's not really a broken invariant if it's a user argument.  When taking user input, we'll want to handle the out of bounds case without panicking.  The panic is there because before it was theoretically impossible (due to invariants) for this check to fail -- that is no longer the case.